contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -262,6 +258,424 @@ contract MixinMatchOrders is
+            doesLeftMakerAssetProfitExist = true;
+            doesRightMakerAssetProfitExist = true;
+        }
+
+        // Calculate amount given to taker in the left order's maker asset if the left spread will be part of the profit.
+        if (doesLeftMakerAssetProfitExist) {
+            matchedFillResults.profitInLeftMakerAsset = _safeSub(
+                matchedFillResults.left.makerAssetFilledAmount,
+                matchedFillResults.right.takerAssetFilledAmount
+            );
+        }
+
+        // Calculate amount given to taker in the right order's maker asset if the right spread will be part of the profit.
+        if (doesRightMakerAssetProfitExist) {
+            matchedFillResults.profitInRightMakerAsset = _safeSub(
+                matchedFillResults.right.makerAssetFilledAmount,
+                matchedFillResults.left.takerAssetFilledAmount
+            );
+        }
+    }
+
+    function _calculateCompleteRightFill(
+        MatchedFillResults memory matchedFillResults,
+        LibOrder.Order memory leftOrder,
+        uint256 rightMakerAssetAmountRemaining,
+        uint256 rightTakerAssetAmountRemaining
+    )
+        internal
+        pure
+    {
+        matchedFillResults.right.makerAssetFilledAmount = rightMakerAssetAmountRemaining;
+        matchedFillResults.right.takerAssetFilledAmount = rightTakerAssetAmountRemaining;
+        matchedFillResults.left.takerAssetFilledAmount = rightMakerAssetAmountRemaining;
+        // Round down to ensure the left maker's exchange rate does not exceed the price specified by the order.
+        // We favor the left maker when the exchange rate must be rounded and the profit is being paid in the
+        // left maker asset.
+        matchedFillResults.left.makerAssetFilledAmount = _safeGetPartialAmountFloor(
+            leftOrder.makerAssetAmount,
+            leftOrder.takerAssetAmount,
+            rightMakerAssetAmountRemaining
+        );
+    }
+
+    /// @dev Match two complementary orders that have a profitable spread.
+    ///      Each order is filled at their respective price point. However, the calculations are
+    ///      carried out as though the orders are both being filled at the right order's price point.
+    ///      The profit made by the left order goes to the taker (who matched the two orders). This
+    ///      function is needed to allow for reentrant order matching (used by `batchMatchOrders` and
+    ///      `batchMatchOrdersWithMaximalFill`).
+    /// @param leftOrder First order to match.
+    /// @param rightOrder Second order to match.
+    /// @param leftSignature Proof that order was created by the left maker.
+    /// @param rightSignature Proof that order was created by the right maker.
+    /// @param shouldMaximallyFillOrders Indicates whether or not the maximal fill matching strategy should be used
+    /// @return matchedFillResults Amounts filled and fees paid by maker and taker of matched orders.
+    function _matchOrders(
```

Comment:
Nit 1: why is this `private` while `_batchMatchOrders` is `internal`? Probably make that `private` as well.